### PR TITLE
Support opening up container and service ports based on agent config YAML

### PR DESCRIPTION
**Pull request by mitali-salvi**
*Issue #, if available:*

*Description of changes:*
1. Add support for opening up service and container ports for CloudWatch Agent based on the otelConfig provided
2. Add support for live-ness probe if health_check extension is configured

*Testing*
1. Unit tests
2. Tested manually
a. Default ports for receivers --
```
    receivers:
      statsd:
    processors:
    exporters:
      debug:
    service:
      pipelines:
        metrics:
          receivers: [ statsd ]
          exporters: [ debug ]
```
Checking agent container ports
```
Containers:
  otc-container:
    Container ID:   containerd://48e9dcff6b061490a7a3b7b8726eaad0ae31eda0d17a261829c7c83056517e83
    Ports:          4315/TCP, 4316/TCP, 2000/TCP, 4322/TCP
    Host Ports:     4315/TCP, 4316/TCP, 2000/TCP, 4322/TCP
```

Checking agent service ports
```
cloudwatch-agent                                         ClusterIP   10.100.15.241    <none>        4315/TCP,4316/TCP,2000/TCP,4322/TCP   110s
cloudwatch-agent-headless                                ClusterIP   None             <none>        4315/TCP,4316/TCP,2000/TCP,4322/TCP   110s
cloudwatch-agent-monitoring                              ClusterIP   10.100.224.40    <none>        8888/TCP                              110s
```

b. Override ports for receivers --
```
    receivers:
      otlp:
        protocols:
          grpc:
            endpoint: "localhost:4322"
    processors:
    exporters:
      debug:
    service:
      pipelines:
        metrics:
          receivers: [ otlp ]
          exporters: [ debug ]
```
Verified container ports and service ports are updated

c. Conflicting ports for receivers --
```
    receivers:
      otlp:
        protocols:
          grpc:
            endpoint: "localhost:4315" //conflicts with AppSignals
    processors:
    exporters:
      debug:
    service:
      pipelines:
        metrics:
          receivers: [ otlp ]
          exporters: [ debug ]
```
In this case, the JSON config ports are prioritized and duplicate ports from the YAML config are dropped.
Agent will also fail in this case with
```
2024-09-27T18:54:59Z E! Error running agent: cannot start pipelines: listen tcp 127.0.0.1:4315: bind: address already in use 
```
d. Enabling health_check extension should create a liveness probe
```
receivers:
  statsd:
processors:
exporters:
  debug:
extensions:
  zpages:
  health_check:
  pprof:
service:
  extensions: [health_check, pprof, zpages]
  pipelines:
    metrics:
      receivers: [ statsd ]
      exporters: [ debug ]
```

Will create a liveness probe on the agent container
```
Liveness:  http-get http://:13133/ delay=0s timeout=1s period=10s #success=1 #failure=3
```
Although this scenario only works when the agent runs on a non sidecar mode - https://github.com/open-telemetry/opentelemetry-operator/issues/571 and mounted on the application pod.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by musa-asad — APPROVED** [review]

File: internal/manifests/collector/adapters/config_to_ports.go [hunk 1]
```diff
@@ -58,3 +65,89 @@ func ConfigToMetricsPort(logger logr.Logger, config map[interface{}]interface{})
...
+		return ports[i].Name < ports[j].Name
+	})
+
+	return ports, nil
+}
+
+// ConfigToComponentPorts converts the incoming configuration object into a set of service ports required by the exporters.
+func ConfigToComponentPorts(logger logr.Logger, cType ComponentType, config map[interface{}]interface{}) ([]corev1.ServicePort, error) {
+	// now, we gather which ports we might need to open
+	// for that, we get all the exporters and check their `endpoint` properties,
```
Comment (sky333999) [inline internal/manifests/collector/adapters/config_to_ports.go:86]: Why do we need to open ports for exporters? Id expect we only listen on ports for receivers.

Comment (mitali-salvi) [reply internal/manifests/collector/adapters/config_to_ports.go:86]: Ports only opened for receivers. Updated the comment to reflect the same

File: internal/manifests/collector/ports.go [hunk 2]
```diff
@@ -260,3 +279,10 @@ func portFromEndpoint(endpoint string) (int32, error) {
 
 	return int32(port), err
 }
+
+func isDuplicatePort(portsMap map[string]corev1.ContainerPort, servicePort corev1.ServicePort) bool {
+	for _, containerPort := range portsMap {
+		return containerPort.Protocol == servicePort.Protocol && containerPort.ContainerPort == servicePort.Port
```
Comment (jefchien) [inline internal/manifests/collector/ports.go:285]: Isn't this going to just return for the first port in the map? This doesn't seem like it's going to search the map.

File: internal/manifests/collector/adapters/config_to_ports.go [hunk 3]
```diff
@@ -58,3 +65,86 @@ func ConfigToMetricsPort(logger logr.Logger, config map[interface{}]interface{})
...
+
+	components, ok := componentsProperty.(map[interface{}]interface{})
+	if !ok {
+		return nil, fmt.Errorf("%ss doesn't contain valid components", cType.String())
+	}
+
+	compEnabled := getEnabledComponents(config, cType)
+
+	if compEnabled == nil {
+		return nil, fmt.Errorf("no enabled %ss available as part of the configuration", cType)
```
Comment (jefchien) [inline internal/manifests/collector/adapters/config_to_ports.go:109]: This doesn't seem like it should necessarily emit an error. It's valid to have a config with only scraping receivers.

File: internal/manifests/collector/parser/receiver/receiver.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+// IsRegistered checks whether a parser is registered with the given name.
+func IsRegistered(name string) bool {
+	_, ok := registry[name]
+	return ok
+}
+
+var (
+	endpointKey      = "endpoint"
+	listenAddressKey = "listen_address"
+	scraperReceivers = map[string]struct{}{
```
Comment (jefchien) [inline internal/manifests/collector/parser/receiver/receiver.go:54]: nit: Can we narrow this down to what we actually support (e.g. prometheus)?

File: internal/manifests/collector/parser/receiver/receiver.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+func singlePortFromConfigEndpoint(logger logr.Logger, name string, config map[interface{}]interface{}) *v1.ServicePort {
+	var endpoint interface{}
+	switch {
+	// syslog receiver contains the endpoint
+	// that needs to be exposed one level down inside config
+	// i.e. either in tcp or udp section with field key
+	// as `listen_address`
+	case name == "syslog":
```
Comment (jefchien) [inline internal/manifests/collector/parser/receiver/receiver.go:100]: nit: Unsupported receiver

File: internal/manifests/collector/parser/receiver/receiver_collectd.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"github.com/go-logr/logr"
+
+	"github.com/aws/amazon-cloudwatch-agent-operator/internal/manifests/collector/parser"
+)
+
+const parserNameCollectd = "__collectd"
+
+// NewCollectdReceiverParser builds a new parser for Collectd receivers, from the contrib repository.
+func NewCollectdReceiverParser(logger logr.Logger, name string, config map[interface{}]interface{}) parser.ComponentPortParser {
```
Comment (jefchien) [inline internal/manifests/collector/parser/receiver/receiver_collectd.go:15]: nit: We aren't supporting this receiver.

File: internal/manifests/collector/adapters/config_to_ports.go [hunk 7]
```diff
@@ -4,19 +4,26 @@
...
 import (
+	"fmt"
 	"net"
+	"sort"
 	"strconv"
 	"strings"
 
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/aws/amazon-cloudwatch-agent-operator/internal/manifests/collector/parser"
```
[the remark this answers is not in the document]
Comment (mitali-salvi) [reply internal/manifests/collector/adapters/config_to_ports.go]: updated

File: internal/manifests/collector/adapters/config_to_ports.go [hunk 8]
```diff
@@ -58,3 +65,86 @@ func ConfigToMetricsPort(logger logr.Logger, config map[interface{}]interface{})
...
 }
+
+func GetServicePortsFromCWAgentOtelConfig(logger logr.Logger, config map[interface{}]interface{}) ([]corev1.ServicePort, error) {
+	ports, err := ConfigToComponentPorts(logger, ComponentTypeReceiver, config)
+	if err != nil {
+		logger.Error(err, "there was a problem while getting the ports from the receivers")
+		return nil, err
+	}
+
+	sort.Slice(ports, func(i, j int) bool {
```
[the remark this answers is not in the document]
Comment (mitali-salvi) [reply internal/manifests/collector/adapters/config_to_ports.go]: removed

File: internal/manifests/collector/ports.go (hunk 2 above)
Comment (mitali-salvi) [reply internal/manifests/collector/ports.go]: good catch ! updated

File: internal/manifests/collector/parser/receiver/receiver.go (hunk 4 above)
Comment (mitali-salvi) [reply internal/manifests/collector/parser/receiver/receiver.go:54]: updated to only supported receivers

File: internal/manifests/collector/parser/receiver/receiver_collectd.go (hunk 6 above)
Comment (mitali-salvi) [reply internal/manifests/collector/parser/receiver/receiver_collectd.go]: removed

File: internal/manifests/collector/adapters/config_to_ports.go (hunk 1 above)
Comment (jefchien) [reply internal/manifests/collector/adapters/config_to_ports.go:80]: nit: The comments for and in this function still refer to exporters.

File: internal/manifests/collector/ports_test.go [hunk 9]
```diff
@@ -197,14 +197,67 @@ func TestSpecPortsOverrideGetContainerPorts(t *testing.T) {
...
+	assert.Equal(t, int32(4315), containerPorts[CWA+AppSignalsGrpc].ContainerPort)
+	assert.Equal(t, CWA+AppSignalsGrpc, containerPorts[CWA+AppSignalsGrpc].Name)
+	assert.Equal(t, int32(4316), containerPorts[CWA+AppSignalsHttp].ContainerPort)
+	assert.Equal(t, CWA+AppSignalsHttp, containerPorts[CWA+AppSignalsHttp].Name)
+	assert.Equal(t, int32(2000), containerPorts[CWA+AppSignalsProxy].ContainerPort)
+	assert.Equal(t, CWA+AppSignalsProxy, containerPorts[CWA+AppSignalsProxy].Name)
+}
+
+func TestIsDuplicatePort(t *testing.T) {
+	containerPorts := map[string]corev1.ContainerPort{
```
Comment (jefchien) [inline internal/manifests/collector/ports_test.go:249]: nit: Add multiple ports so we can verify that the search/dedup works.

File: internal/manifests/collector/adapters/config_to_ports.go [hunk 10]
```diff
@@ -58,3 +63,82 @@ func ConfigToMetricsPort(logger logr.Logger, config map[interface{}]interface{})
...
+		cmptName := key.(string)
+		var cmptParser parser.ComponentPortParser
+		var err error
+		cmptParser, err = receiverParser.For(logger, cmptName, extractedComponent)
+		if err != nil {
+			logger.V(2).Info("no parser found for '%s'", cmptName)
+			continue
+		}
+
+		exprtPorts, err := cmptParser.Ports()
```
Comment (jefchien) [inline internal/manifests/collector/adapters/config_to_ports.go:128]: nit: 
```suggestion
		rcvrPorts, err := cmptParser.Ports()
```

**Review by jefchien — APPROVED** [review]

**Pull request merged**